index.bs

```diff
@@ -124,50 +124,97 @@ Issue: We should consider having further normative restrictions on file names th
+and <dfn for="file entry">pending steps</dfn> (a [=/list=] of sets of steps).
 
 A user agent has an associated <dfn>file system queue</dfn> which is the
 result of [=starting a new parallel queue=]. This queue is to be used for all
 file system operations.
 
 <div algorithm>
-To <dfn for="file entry/lock">take</dfn> a [=file entry/lock=] with a |value| of
-"`exclusive`" or "`shared`" on a given [=file entry=] |file|:
-
-1. Let |lock| be the |file|'s [=file entry/lock=].
-1. Let |count| be the |file|'s [=file entry/shared lock count=].
-1. If |value| is "`exclusive`":
-   1. If |lock| is "`open`":
-     1. Set lock to "`taken-exclusive`".
-     1. Return "`success`".
-1. If |value| is "`shared`":
-   1. If |lock| is "`open`":
-     1. Set |lock| to "`taken-shared`".
-     1. Set |count| to 1.
-     1. Return "`success`".
-   1. Otherwise, if |lock| is "`taken-shared`":
-     1. Increase |count| by 1.
-     1. Return "`success`".
-1. Return "`failure`".
+To <dfn for="file entry" id=file-entry-lock-take>take a lock</dfn> on a given [=file entry=] |file|
+with a |lockType| (a [=lock type=]), a set of steps |resultSteps|, a [=realm/global object=] |global|,
+and a |lockResult| (a [=string=]):
+
+1. [=Assert=]: |lockType| is not "`open`".
+1. Let |pendingLock| be the |file|'s [=file entry/pending lock=].
+1. Let |pendingSteps| be the |file|'s [=file entry/pending steps=].
+1. Let |activeLock| be the |file|'s [=file entry/lock=].
+1. If |pendingLock|'s [=file lock/type=] is not "`open`":
+  1. Set |activeLock| to |pendingLock|.
+1. Let |activeLockType| be the |activeLock|'s [=file lock/type=].
+1. If |activeLockType| is not "`open`":
+  1. If |activeLockType| is not equal to |lockType| or is "`exclusive`":
+    1. Let |evictedStatus| be the result of [=file lock/evicting all globals=] on |activeLock|.
+    1. If |evictedStatus| is "`failure`":
+      1. Set |lockResult| to `failure`.
+      1. Run |resultSteps|.
+      1. Return.
+    1. [=list/Empty=] |pendingLock|'s [=file lock/globals=].
+    1. [=list/Empty=] |pendingSteps|.
+    1. Set |activeLock| to |pendingLock|.
+1. Set |lockResult| to `success`.
+1. Set the |activeLock|'s [=file lock/type=] to |lockType|.
+1. [=list/Append=] |global| to |activeLock|'s [=file lock/globals=].
+1. If |activeLock| is the |pendingLock|:
+  1. [=list/Append=] |resultSteps| to |pendingSteps|.
+1. Otherwise:
+  1. Run |resultSteps|.
 
 Note: These steps have to be run on the [=file system queue=].
 
 </div>
 
+<div algorithm>
+To <dfn for="file lock">evict all globals</dfn> for a [=file entry/lock=] |lock|.
+
+1. [=set/For each=] |global| of |lock|'s [=file lock/globals=]:
+  1. If |global| is a {{Window}} object whose [=associated Document=] is [=Document/fully active=]:
+    1. Return "`failure`".
+  1. If |global| is a {{WorkerGlobalScope}} object whose [=WorkerGlobalScope/closing=] flag is false and whose {{worker}} is not a [[html/workers#the-worker's-lifetime|suspendable worker]]:
```

Comment:
Uncertain if this is the right condition to exclude workers. This is taken from another [PR](https://github.com/whatwg/html/pull/8972) with BFCache interaction.